copy.php

```diff
@@ -355,152 +355,171 @@ function ReloadAndDisplay($oPage, $oObj, $sMessageId = '', $sMessage = '', $sSev
+				} else {
+					$oAlert= AlertUIBlockFactory::MakeForFailure(Dict::S('UI:Error:ObjectAlreadyCreated'));
+					$oAlert->SetIsCollapsible(false);
+					$oP->AddSubBlock($oAlert);
+				}
 			}
-			$sStateAttCode = MetaModel::GetStateAttributeCode($sClass);
-			if (!empty($sStateAttCode))
+			else
 			{
-				$sTargetState = utils::ReadPostedParam('obj_state', '');
-				if ($sTargetState != '')
+				$oObj = MetaModel::NewObject($sClass);
+				try
+				{
+					// Do not write attributes that will be updated with posted data (conflict with case logs)
+					iTopObjectCopier::PrepareObject($aRuleData, $oObj, $oSourceObject, true);
+				}
+				catch (Exception $e)
 				{
-					$oObj->Set($sStateAttCode, $sTargetState);
+					iTopObjectCopier::LogError($iRule, 'preset - '.$e->getMessage());
+				}
+				$sStateAttCode = MetaModel::GetStateAttributeCode($sClass);
+				if (!empty($sStateAttCode))
+				{
+					$sTargetState = utils::ReadPostedParam('obj_state', '');
+					if ($sTargetState != '')
+					{
+						$oObj->Set($sStateAttCode, $sTargetState);
+					}
 				}
+				$oObj->UpdateObjectFromPostedForm();
 			}
-			$oObj->UpdateObjectFromPostedForm();
-		}
-		if (isset($oObj) && is_object($oObj))
-		{
-			$sClass = get_class($oObj);
-			$sClassLabel = MetaModel::GetName($sClass);
-
-			list($bRes, $aIssues) = $oObj->CheckToWrite();
-			if ($bRes)
+			if (isset($oObj) && is_object($oObj))
 			{
-				$oObj->DBInsert();
+				$sClass = get_class($oObj);
+				$sClassLabel = MetaModel::GetName($sClass);
 
-				// Specific to itop-object-copier
-				// Note: must be done when the id is known
-				try
+				$sStateAttCode = MetaModel::GetStateAttributeCode($sClass);
+				$bRes = true;
+				if (!empty($sStateAttCode))
 				{
-					$oSourceObject = MetaModel::GetObject($sSourceClass, $iSourceId);
-					iTopObjectCopier::RetrofitOnSourceObject($aRuleData, $oObj, $oSourceObject);
-					$oSourceObject->DBUpdate();
-
-					$sMessage = iTopObjectCopier::FormatMessage($aRuleData, 'report_label', $oSourceObject);
-					cmdbAbstractObject::SetSessionMessage(get_class($oObj), $oObj->GetKey(), 'object-copier', $sMessage, 'info', 0, true /* must not exist */);
+					$sTargetState = utils::ReadPostedParam('obj_state', '');
+					if ($sTargetState != '')
+					{
+						$sOrigState = utils::ReadPostedParam('obj_state_orig', '');
+						if ($sTargetState != $sOrigState)
+						{
+							$aIssues[] = Dict::S('UI:StateChanged');
+							$bRes = false;
+						}
+						$oObj->Set($sStateAttCode, $sTargetState);
+					}
 				}
-				catch (Exception $e)
+				if($bRes)
 				{
-					iTopObjectCopier::LogError($iRule, 'retrofit - '.$e->getMessage());
-					$sMessage = Dict::Format('object-copier:error:retrofit', $e->getMessage());
-					cmdbAbstractObject::SetSessionMessage(get_class($oObj), $oObj->GetKey(), 'object-copier', $sMessage, 'error', 0, true /* must not exist */);
+					list($bRes, $aIssues) = $oObj->CheckToWrite();
 				}
+				if ($bRes)
```

Comment:
2 times `$bRes` used close to each other is confusing. Can the new one be renamed to be clearer like `$bStateChanged`?

Otherwise the `CheckToWrite` could be moved to an else block for `if ($sTargetState != $sOrigState)`.